### PR TITLE
ImageField: Stop parsing images using Regex

### LINKED ISSUE #5874: [ACRA 2.9] BasicImageFieldController - Image Regex Failure
Not a crash bug, but it triggers an exception handler added to diagnose the underlying cause of #5513 through an unexpected path.

This isn't the cause of #5513, but it'll add noise to ACRA, and it's a bug that we can fix

Repro steps:

1. Add a note with a field: `'/>`
2. Save
3. Edit Note
4. Click Paperclip

----

https://couchdb.ankidroid.org/acralyzer/_design/acralyzer/index.html#/report-details/72f0661e-6331-475f-8c0f-3c69ef456531

```
java.io.FileNotFoundException: /sdcard/AnkiDroid/collection.media/paste-12893491822593.jpg" /><div>&nbsp; &nbsp;-·-<span class="Apple-tab-span" style="white-space:pre: open failed: ENOENT (No such file or directory)
at libcore.io.IoBridge.open(IoBridge.java:496)
at java.io.FileInputStream.<init>(FileInputStream.java:159)
at com.ichi2.utils.BitmapUtil.decodeFile(BitmapUtil.java:3)
at com.ichi2.anki.multimediacard.fields.BasicImageFieldController.setPreviewImage(BasicImageFieldController.java:3)
at com.ichi2.anki.multimediacard.fields.BasicImageFieldController.createUI(BasicImageFieldController.java:3)
at com.ichi2.anki.multimediacard.activity.MultimediaEditFieldActivity.recreateEditingUi(MultimediaEditFieldActivity.java:15)
at com.ichi2.anki.multimediacard.activity.MultimediaEditFieldActivity.onCreate(MultimediaEditFieldActivity.java:11)
at android.app.Activity.performCreate(Activity.java:8086)
at android.app.Activity.performCreate(Activity.java:8074)
at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1313)
at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3755)
at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3961)
at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:91)
at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:149)
at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:103)
at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2386)
at android.os.Handler.dispatchMessage(Handler.java:107)
at android.os.Looper.loop(Looper.java:213)
at android.app.ActivityThread.main(ActivityThread.java:8178)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:513)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1101)
Caused by: android.system.ErrnoException: open failed: ENOENT (No such file or directory)
at libcore.io.Linux.open(Native Method)
at libcore.io.ForwardingOs.open(ForwardingOs.java:167)
at libcore.io.BlockGuardOs.open(BlockGuardOs.java:252)
at libcore.io.ForwardingOs.open(ForwardingOs.java:167)
at android.app.ActivityThread$AndroidOs.open(ActivityThread.java:8046)
at libcore.io.IoBridge.open(IoBridge.java:482)
... 21 more
```

<details><summary>Code Refs</summary>

Exception handler
https://github.com/ankidroid/Anki-Android/blob/2353c84b3a23c3f82c514b87c7314fb472aa38c0/AnkiDroid/src/main/java/com/ichi2/utils/BitmapUtil.java#L75-L78

Parsing code:
https://github.com/ankidroid/Anki-Android/blob/2353c84b3a23c3f82c514b87c7314fb472aa38c0/AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/ImageField.java#L143-L153

Assumption we have an image field:
https://github.com/ankidroid/Anki-Android/blob/2353c84b3a23c3f82c514b87c7314fb472aa38c0/AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/NoteService.java#L79-L80

No assumption that the file is invalid:
https://github.com/ankidroid/Anki-Android/blob/2353c84b3a23c3f82c514b87c7314fb472aa38c0/AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java#L281-L290

</details>

### PR DESCRIPTION
**Pull request by david-allison-1**
## Purpose / Description

* The image `src` was not being extracted correctly
* Passing in an invalid image to ImageFieldController started a preview, throwing a FileNotFound Exception, which was logged by ACRA
* Passing a null image to BitmapUtil threw an exception, now returns null

## Fixes
Fixes #5874

## Approach

* We use JSoup instead of Regex
* We avoid logging to ACRA on a regular `FileNotFound`

## How Has This Been Tested?

Unit tests, now displays images on the test data

## Checklist
- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/utils/ExifUtil.java [hunk 1]
```diff
@@ -21,12 +21,15 @@
...
+
+import androidx.annotation.NonNull;
 import androidx.exifinterface.media.ExifInterface;
 
 import java.io.File;
 
 public class ExifUtil {
-    public static Bitmap rotateFromCamera(File theFile, Bitmap bmp) {
+    @NonNull
+    public static Bitmap rotateFromCamera(File theFile, @NonNull Bitmap bmp) {
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/utils/ExifUtil.java]: Codacy Issue found: [Avoid reassigning parameters such as 'bmp'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5264895)

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/ImageField.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5264895&bid=17247493)

File: AnkiDroid/src/main/java/com/ichi2/utils/ExifUtil.java (hunk 1 above)
Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/utils/ExifUtil.java]: No change in the method

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/ImageField.java [hunk 2]
```diff
@@ -21,10 +21,14 @@
 
 import com.ichi2.libanki.Collection;
 
+import org.jetbrains.annotations.NotNull;
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/ImageField.java]: jetbrains annotation

Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/ImageField.java]: 🤦‍♂

**Comment by mikehardy** [comment]
Honestly didn't realize that was what jsoup was for, despite my reflexive response to all HTML parsing! And it's already baked into the APK, why not use the right tool for the job? These all look good to me pending the annotation change, which is just a nit. Thanks @david-allison-1 

Release-planning-wise I just did my first translation string sync today as maintainer (#5875 ) so we're moving along - should be a 2.9.6 beta very shortly, and a 2.9.6 stable release probably mid-week with the current massive batch of fixes in it, then the focus moves to get 2.10 out the door ASAP (as it should be pretty stable already). I can pick small fixes into a new 2.9.x release but I'll try to avoid it in order to maintain focus as most of the big crashes are fixed now. Then finally we can work through the PR queue for real for a 2.11 - Arthur in particular has been patient on his and I'd like to honor his work by getting it in

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/ImageField.java (hunk 2 above)
Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/ImageField.java]: Any way I can contribute to an auto-review for this? It'll happen again, and It's a bad habit that I don't need to waste your time with

Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/ImageField.java]: might just be a config thing you can change? https://www.jetbrains.com/help/idea/nullable-notnull-configuration.html#

Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/ImageField.java]: 🤦‍♀ Thanks

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/ImageField.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5264895&bid=17247493)

File: AnkiDroid/src/main/java/com/ichi2/utils/ExifUtil.java (hunk 1 above)
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/utils/ExifUtil.java]: Codacy Issue found: [Avoid reassigning parameters such as 'bmp'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5264895)

**Comment by mikehardy** [comment]
out now in 2.10alpha58 as soon as google is finished processing it

**Pull request merged**